### PR TITLE
remove FilterFunc and use SelectionPredicate everywhere

### PR DESCRIPTION
**Pull request by dixudx**
**What this PR does / why we need it**:
> // FilterFunc takes an API object and returns true if the object satisfies some requirements.
// TODO: We will remove this type and use SelectionPredicate everywhere.
type FilterFunc func(obj runtime.Object) bool

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
/assign @liggitt @wojtek-t 
**Release note**:

```release-note
None
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/apis/example/types.go [hunk 1]
```diff
@@ -33,7 +33,7 @@ type (
 type Pod struct {
 	metav1.TypeMeta
 	// +optional
-	metav1.ObjectMeta
+	metav1.ObjectMeta `json:"metadata,omitempty"`
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/apis/example/types.go]: Why do you add this one?

I think we should be consistent - either add json tags to all fields, or to any of those.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd/etcd_helper.go [hunk 2]
```diff
@@ -407,7 +407,7 @@ func (h *etcdHelper) decodeNodeList(nodes []*etcd.Node, filter storage.FilterFun
 			}
 			// being unable to set the version does not prevent the object from being extracted
 			_ = h.versioner.UpdateObject(obj, node.ModifiedIndex)
-			if filter(obj) {
+			if matched, _ := filter.Matches(obj); matched {
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd/etcd_helper.go]: That is potentially changing the semantic in case of error (if function returns true, err, then it will be true, previously being false).

In general, we shouldn't ignore errors (both here an below).

File: staging/src/k8s.io/apiserver/pkg/apis/example/types.go (hunk 1 above)
Comment (dixudx) [reply staging/src/k8s.io/apiserver/pkg/apis/example/types.go]: For test https://github.com/dixudx/kubernetes/blob/b7763c4cce9b4190093fdb346debe53fff39f79f/staging/src/k8s.io/apiserver/pkg/storage/etcd/etcd_watcher_test.go#L132.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd/etcd_watcher.go [hunk 3]
```diff
@@ -105,11 +105,9 @@ const watchWaitDuration = 100 * time.Millisecond
 
 // newEtcdWatcher returns a new etcdWatcher; if list is true, watch sub-nodes.
 // The versioner must be able to handle the objects that transform creates.
-func newEtcdWatcher(
-	list bool, quorum bool, include includeFunc, filter storage.FilterFunc,
+func newEtcdWatcher(list bool, quorum bool, include includeFunc, filter storage.SelectionPredicate,
```
Comment (hzxuzhonghu) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd/etcd_watcher.go]: `filter storage.SelectionPredicate ` name odd

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go [hunk 4]
```diff
@@ -250,14 +250,15 @@ func (wc *watchChan) processEvent(wg *sync.WaitGroup) {
 }
 
 func (wc *watchChan) filter(obj runtime.Object) bool {
-	if wc.internalFilter == nil {
+	if wc.internalFilter.Empty() {
```
Comment (hzxuzhonghu) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go]: I do not think it is better than it was .

Comment (dixudx) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go]: The data type has been changed. It is not a pointer any more.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd/etcd_helper.go (hunk 2 above)
Comment (dixudx) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd/etcd_helper.go]: Good catch. @wojtek-t Thanks for pointing out.

**Comment by liggitt** [comment]
this is post-1.9, right? prioritizing reviews for freeze today

**Comment by dixudx** [comment]
@liggitt Yeah, for post-1.9. Not in a hurry.

**Comment by dixudx** [comment]
code freeze has been lifted.
ping @liggitt @sttts PTAL.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd/etcd_watcher.go [hunk 5]
```diff
@@ -391,7 +389,10 @@ func (w *etcdWatcher) sendModify(res *etcd.Response) {
 		// the resourceVersion to resume will never be able to get past a bad value.
 		return
 	}
-	curObjPasses := w.filter(curObj)
+	curObjPasses, err := w.pred.Matches(curObj)
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd/etcd_watcher.go]: nit: I think it would be more readable to do:
```
curObjPasses := false
if matched, err := w.pred.Matches(curObj); err == nil && matched {
  curObjPasses = true
}
```

[also for consistency]

File: staging/src/k8s.io/apiserver/pkg/storage/etcd/etcd_watcher.go [hunk 6]
```diff
@@ -400,10 +401,13 @@ func (w *etcdWatcher) sendModify(res *etcd.Response) {
 			if err := w.versioner.UpdateObject(oldObj, res.Node.ModifiedIndex); err != nil {
 				utilruntime.HandleError(fmt.Errorf("failure to version api object (%d) %#v: %v", res.Node.ModifiedIndex, oldObj, err))
 			}
-			oldObjPasses = w.filter(oldObj)
+			oldObjPasses, err = w.pred.Matches(oldObj)
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd/etcd_watcher.go]: and same here

**Review by wojtek-t — COMMENTED** [review]
Just two minor nits - other than that LGTM.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd/etcd_watcher.go (hunk 5 above)
Comment (dixudx) [reply staging/src/k8s.io/apiserver/pkg/storage/etcd/etcd_watcher.go]: @wojtek-t Addressed. PTAL. Thanks.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go [hunk 7]
```diff
@@ -111,14 +111,14 @@ func (w *watcher) createWatchChan(ctx context.Context, key string, rev int64, re
...
-		internalFilter:    storage.SimpleFilter(pred),
+		internalPred:      pred,
 		incomingEventChan: make(chan *event, incomingBufSize),
 		resultChan:        make(chan watch.Event, outgoingBufSize),
 		errChan:           make(chan error, 1),
 	}
 	if pred.Empty() {
 		// The filter doesn't filter out any object.
-		wc.internalFilter = nil
+		wc.internalPred = storage.Everything
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/watcher.go]: why is this necessary if the acceptAll() func already returns true based on `Empty()`?
